purchases/src/main/kotlin/com/revenuecat/purchases/PostReceiptHelper.kt

```diff
@@ -0,0 +1,143 @@
+            onError = { error, shouldConsumePurchase, _, _ ->
+                if (shouldConsumePurchase) {
+                    deviceCache.addSuccessfullyPostedToken(purchaseToken)
+                }
+                onError(error)
+            }
+        )
+    }
+
+    /**
+     * This method will post a StoreTransaction and optionally a StoreProduct info to the backend.
+     * It will consume the purchase if finishTransactions is true.
+     */
+    fun postTransactionToBackend(
+        purchase: StoreTransaction,
+        storeProduct: StoreProduct?,
+        isRestore: Boolean,
+        appUserID: String,
+        onSuccess: (SuccessfulPurchaseCallback)? = null,
+        onError: (ErrorPurchaseCallback)? = null
+    ) {
+        val receiptInfo = ReceiptInfo(
+            productIDs = purchase.productIds,
+            offeringIdentifier = purchase.presentedOfferingIdentifier,
+            storeProduct = storeProduct,
+            subscriptionOptionId = purchase.subscriptionOptionId
+        )
+        syncSubscriberAttributesAndPostReceipt(
+            appUserID = appUserID,
+            purchaseToken = purchase.purchaseToken,
+            isRestore = isRestore,
+            receiptInfo = receiptInfo,
+            storeUserID = purchase.storeUserID,
+            marketplace = purchase.marketplace,
+            onSuccess = { info, _ ->
+                billing.consumeAndSave(finishTransactions, purchase)
+                onSuccess?.let { it(purchase, info) }
+            },
+            onError = { error, shouldConsumePurchase, _, _ ->
+                if (shouldConsumePurchase) {
+                    billing.consumeAndSave(finishTransactions, purchase)
+                }
+                onError?.let { it(purchase, error) }
+            }
+        )
+    }
+
+    private fun syncSubscriberAttributesAndPostReceipt(
+        appUserID: String,
+        purchaseToken: String,
+        isRestore: Boolean,
+        receiptInfo: ReceiptInfo,
+        storeUserID: String?,
+        marketplace: String?,
+        onSuccess: PostReceiptDataSuccessCallback,
+        onError: PostReceiptDataErrorCallback
+    ) {
+        subscriberAttributesManager.getUnsyncedSubscriberAttributes(appUserID) { unsyncedSubscriberAttributesByKey ->
+            backend.postReceiptData(
+                purchaseToken = purchaseToken,
+                appUserID = appUserID,
+                isRestore = isRestore,
+                observerMode = !finishTransactions,
+                subscriberAttributes = unsyncedSubscriberAttributesByKey.toBackendMap(),
+                receiptInfo = receiptInfo,
+                storeAppUserID = storeUserID,
+                marketplace = marketplace,
+                onSuccess = { customerInfo, responseBody ->
+                    subscriberAttributesManager.markAsSynced(
+                        appUserID,
+                        unsyncedSubscriberAttributesByKey,
+                        responseBody.getAttributeErrors()
```

Comment:
how does the subs attrs manager know which ones we did sync in this post? 

I don't see any locking, so in theory you could have added unsynced attrs between calling this method and getting the response back, and we're only sending in the unsynced ones. So could the subs attrs manager be marking new attributes as synced when they were added after line 110 executed but before this completion got called?

Reply:
No, note how on line 110 we get the `unsyncedSubscriberAttributesByKey` and those are the ones we `markAsSynced`. So if more attributes are added in between, they will be sent on the next opportunity. Lmk if you meant something else!